File: app/src/main/java/com/mobile/meplayer/data/M3uParser.kt
package com.mobile.meplayer.data

/**
 * M3U/M3U8 解析器，对应 Flutter 的 `m3u_parser.dart`。
 */
data class M3uChannel(
    val name: String,
    val url: String,
    val group: String = "未分组",
    val logo: String = "",
    val id: String = ""
)

object M3uParser {

    private val attrRegex = Regex("""(\w[\w-]*)="([^"]*)"""", RegexOption.IGNORE_CASE)

    fun parse(content: String): List<M3uChannel> {
        val lines = content.split(Regex("\r?\n"))
        val channels = mutableListOf<M3uChannel>()

        var name = ""
        var group = "未分组"
        var logo = ""
        var id = ""

        for (rawLine in lines) {
            val line = rawLine.trim()
            if (line.isEmpty() || line == "#EXTM3U") continue

            if (line.startsWith("#EXTINF:")) {
                name = attr(line, "tvg-name") ?: attr(line, "title") ?: commaName(line)
                group = attr(line, "group-title") ?: "未分组"
                logo = attr(line, "tvg-logo") ?: ""
                id = attr(line, "tvg-id") ?: ""
            } else if (!line.startsWith("#")) {
                channels.add(
                    M3uChannel(
                        name = if (name.isEmpty()) guessName(line) else name,
                        url = line,
                        group = group,
                        logo = logo,
                        id = id
                    )
                )
                name = ""
                group = "未分组"
                logo = ""
                id = ""
            }
        }
        return channels
    }

    private fun attr(line: String, key: String): String? {
        val pattern = Regex("""$key="([^"]*)"""", RegexOption.IGNORE_CASE)
        return pattern.find(line)?.groupValues?.get(1)?.trim()
    }

    private fun commaName(line: String): String {
        val idx = line.lastIndexOf(',')
        return if (idx < 0) "" else line.substring(idx + 1).trim()
    }

    private fun guessName(url: String): String {
        return runCatching {
            val uri = android.net.Uri.parse(url)
            val segments = uri.pathSegments
            val last = segments.lastOrNull() ?: return url
            last.substringBeforeLast('.', last)
        }.getOrDefault(url)
    }

    /** 按分组组织频道，保持插入顺序 */
    fun groupBy(channels: List<M3uChannel>): Map<String, List<M3uChannel>> {
        val map = LinkedHashMap<String, MutableList<M3uChannel>>()
        for (ch in channels) {
            map.getOrPut(ch.group) { mutableListOf() }.add(ch)
        }
        return map
    }
}

File: Services/M3uParser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace MEPlayer.Services;

/// <summary>简单 M3U/M3U8 解析器，与 Flutter 端 m3u_parser.dart 完全一致。</summary>
public sealed class M3uChannel
{
    public string Name { get; set; } = "";
    public string Url { get; set; } = "";
    public string Group { get; set; } = "未分组";
    public string Logo { get; set; } = "";
    public string Id { get; set; } = "";
}

public static class M3uParser
{
    private static readonly Regex AttrRe = new(@"""([^""]*)""", RegexOptions.Compiled);

    public static List<M3uChannel> Parse(string content)
    {
        var channels = new List<M3uChannel>();
        var lines = content.Split(new[] { "\r\n", "\n" }, StringSplitOptions.None);

        string name = "", group = "未分组", logo = "", id = "";

        foreach (var raw in lines)
        {
            var line = raw.Trim();
            if (string.IsNullOrEmpty(line) || line == "#EXTM3U") continue;

            if (line.StartsWith("#EXTINF:"))
            {
                name = Attr(line, "tvg-name") ?? Attr(line, "title") ?? CommaName(line);
                group = Attr(line, "group-title") ?? "未分组";
                logo = Attr(line, "tvg-logo") ?? "";
                id = Attr(line, "tvg-id") ?? "";
            }
            else if (!line.StartsWith("#"))
            {
                channels.Add(new M3uChannel
                {
                    Name = string.IsNullOrEmpty(name) ? GuessName(line) : name,
                    Url = line,
                    Group = group,
                    Logo = logo,
                    Id = id,
                });
                name = ""; group = "未分组"; logo = ""; id = "";
            }
        }
        return channels;
    }

    private static string? Attr(string line, string key)
    {
        var idx = line.IndexOf(key + "=\"", StringComparison.OrdinalIgnoreCase);
        if (idx < 0) return null;
        idx += key.Length + 2;
        var end = line.IndexOf('"', idx);
        if (end < 0) return null;
        return line.Substring(idx, end - idx).Trim();
    }

    private static string CommaName(string line)
    {
        var idx = line.LastIndexOf(',');
        return idx < 0 ? "" : line.Substring(idx + 1).Trim();
    }

    private static string GuessName(string url)
    {
        try
        {
            var uri = new Uri(url);
            var seg = uri.Segments.Length > 0 ? uri.Segments[^1] : url;
            return seg.Split('.')[0];
        }
        catch { return url; }
    }

    /// <summary>按 group-title 分组，保留插入顺序。</summary>
    public static Dictionary<string, List<M3uChannel>> GroupBy(List<M3uChannel> channels)
    {
        var map = new Dictionary<string, List<M3uChannel>>();
        foreach (var ch in channels)
        {
            if (!map.TryGetValue(ch.Group, out var list))
            {
                list = new List<M3uChannel>();
                map[ch.Group] = list;
            }
            list.Add(ch);
        }
        return map;
    }
}
